services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/CrtResponseFileResponseTransformer.java

```diff
@@ -0,0 +1,110 @@
+ */
+
+package software.amazon.awssdk.services.s3.internal.crt;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.core.async.SdkPublisher;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * When the CRT Response File option is used in a request, the body is streamed directly to the file.
+ * The S3CrtResponseHandlerAdapter in this case will never receive a response body but will call onStream
+ * when the request is complete with a publisher that will complete immediately.
+ * This transformer is effectively a no-op transformer that waits for the stream to complete and then
+ * completes the future with the response.
+ *
+ * @param <ResponseT> Pojo response type.
+ */
+@SdkInternalApi
+public final class CrtResponseFileResponseTransformer<ResponseT> implements AsyncResponseTransformer<ResponseT,
+    ResponseT> {
+
+    private static final Logger log = Logger.loggerFor(CrtResponseFileResponseTransformer.class);
+
+    private volatile CompletableFuture<Void> cf;
+    private volatile ResponseT response;
+
+    @Override
+    public CompletableFuture<ResponseT> prepare() {
+        cf = new CompletableFuture<>();
+        return cf.thenApply(ignored -> response);
+    }
+
+    @Override
+    public void onResponse(ResponseT response) {
+        this.response = response;
+    }
+
+    @Override
+    public void onStream(SdkPublisher<ByteBuffer> publisher) {
+        publisher.subscribe(new OnCompleteSubscriber(cf, this::exceptionOccurred));
+    }
+
+    @Override
+    public void exceptionOccurred(Throwable throwable) {
+        if (cf != null) {
+            cf.completeExceptionally(throwable);
+        } else {
+            log.warn(() -> "An exception occurred before the call to prepare() was able to instantiate the CompletableFuture. "
+                           + "The future cannot be completed exceptionally because it is null");
+        }
+    }
+
+    private static final class OnCompleteSubscriber implements Subscriber<ByteBuffer> {
+
+        private final CompletableFuture<Void> future;
+        private final Consumer<Throwable> onErrorMethod;
+        private Subscription subscription;
+
+        private OnCompleteSubscriber(CompletableFuture<Void> future, Consumer<Throwable> onErrorMethod) {
+            this.future = future;
+            this.onErrorMethod = onErrorMethod;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (this.subscription != null) {
+                s.cancel();
+                return;
+            }
+            this.subscription = s;
+            // do not request data from the subscription since body is written directly to file
```

Comment:
Should we still request data to be compliant to the reactive streams spec? Hopefully the publisher should be able to handle it?

Reply:
On second thought. NVM